modules/core/src/main/java/org/apache/ignite/internal/cdc/ChangeDataCapture.java

```diff
@@ -0,0 +1,531 @@
+            log.info("ConsistentId -\t" + cfg.getConsistentId());
+        }
+
+        PdsFolderSettings<ChangeDataCaptureFileLockHolder> settings =
+            new PdsFolderResolver<>(cfg, log, null, this::tryLock).resolve();
+
+        if (settings == null)
+            throw new RuntimeException("Can't find PDS folder!");
+
+        ChangeDataCaptureFileLockHolder lock = settings.getLockedFileLockHolder();
+
+        if (lock == null) {
+            File consIdDir = new File(settings.persistentStoreRootPath(), settings.folderName());
+
+            lock = tryLock(consIdDir);
+
+            if (lock == null)
+                throw new RuntimeException("Can't lock CDC dir " + settings.consistentId());
+        }
+
+        try {
+            init();
+
+            if (log.isInfoEnabled()) {
+                log.info("CDC dir     -\t" + cdcDir);
+                log.info("Binary meta -\t" + binaryMeta);
+                log.info("Marshaller  -\t" + marshaller);
+                log.info("--------------------------------");
+            }
+
+            state = new ChangeDataCaptureConsumerState(cdcDir.resolve(STATE_DIR));
+
+            initState = state.load();
+
+            if (initState != null && log.isInfoEnabled())
+                log.info("Loaded initial state[state=" + initState + ']');
+
+            consumer.start(log);
+
+            try {
+                consumeWalSegmentsUntilStopped();
+            }
+            finally {
+                consumer.stop();
+
+                if (log.isInfoEnabled())
+                    log.info("Ignite CDC Application stoped.");
+            }
+        }
+        finally {
+            U.closeQuiet(lock);
+        }
+    }
+
+    /** Searches required directories. */
+    private void init() throws IOException, IgniteCheckedException {
+        String consIdDir = cdcDir.getName(cdcDir.getNameCount() - 1).toString();
+
+        Files.createDirectories(cdcDir.resolve(STATE_DIR));
+
+        binaryMeta = CacheObjectBinaryProcessorImpl.binaryWorkDir(cfg.getWorkDirectory(), consIdDir);
+
+        marshaller = MarshallerContextImpl.mappingFileStoreWorkDir(cfg.getWorkDirectory());
+
+        if (log.isDebugEnabled()) {
+            log.debug("Using BinaryMeta directory[dir=" + binaryMeta + ']');
+            log.debug("Using Marshaller directory[dir=" + marshaller + ']');
+        }
+
+        injectResources(consumer.getCdcConsumer());
+    }
+
+    /** Waits and consumes new WAL segments until stoped. */
+    public void consumeWalSegmentsUntilStopped() {
+        try {
+            Set<Path> seen = new HashSet<>();
+
+            long[] lastSgmnt = new long[] {-1};
+
+            while (!stopped) {
+                try (Stream<Path> cdcFiles = Files.walk(cdcDir, 1)) {
+                    Set<Path> exists = new HashSet<>();
+
+                    cdcFiles
+                        .peek(exists::add) // Store files that exists in cdc dir.
```

Comment:
Not sure I got the profit of seen/exists sets.
Both always empty on iteration finish.
And have the same content before `consumeSegment` call.
All you need is to consume every file available with an assertion that index increments?

Have I missed something?

Reply:
Say, we have the following segments on the disk: `1.bin`, `2.bin`, `3.bin`.
Consumer process them but not commit state.
We can't delete those segments from the disk but shouldn't process them twice.

So `seen` set saves what segments we process already but they still exist.